File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/exception/CqlException.java
package org.opencds.cqf.cql.engine.exception;

import org.opencds.cqf.cql.engine.debug.SourceLocator;

public class CqlException extends RuntimeException {
    private static final long serialVersionUID = 1L;

    public CqlException(String message) {
        this(message, null, null, null);
    }

    public CqlException(String message, Throwable cause) {
        this(message, cause, null);
    }

    public CqlException(Throwable cause) {
        this(null, cause, null);
    }

    public CqlException(String message, SourceLocator sourceLocator) {
        this(message, null, sourceLocator);
    }

    public CqlException(String message, Throwable cause, SourceLocator sourceLocator) {
        this(message, cause, sourceLocator, null);
    }

    public CqlException(Throwable cause, SourceLocator sourceLocator) {
        this(null, cause, sourceLocator, null);
    }

    public CqlException(String message, SourceLocator sourceLocator, Severity severity) {
        this(message, null, sourceLocator, severity);
    }

    public CqlException(String message, Throwable cause, SourceLocator sourceLocator, Severity severity) {
        super(message, cause);
        this.sourceLocator = sourceLocator;
        this.severity = severity != null ? severity : Severity.ERROR;
    }

    private final Severity severity;

    public Severity getSeverity() {
        return severity;
    }

    private transient SourceLocator sourceLocator;

    public SourceLocator getSourceLocator() {
        return sourceLocator;
    }

    public void setSourceLocator(SourceLocator sourceLocator) {
        this.sourceLocator = sourceLocator;
    }

    private final Backtrace backtrace = new Backtrace();

    public Backtrace getBacktrace() {
        return this.backtrace;
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/exception/CqlException.kt
package org.opencds.cqf.cql.engine.exception

import org.opencds.cqf.cql.engine.debug.SourceLocator

open class CqlException
@JvmOverloads
constructor(
    message: String?,
    cause: Throwable? = null,
    sourceLocator: SourceLocator? = null,
    severity: Severity? = null,
) : RuntimeException(message, cause) {
    constructor(cause: Throwable?) : this(null, cause, null)

    constructor(
        message: String?,
        sourceLocator: SourceLocator?,
    ) : this(message, null, sourceLocator)

    constructor(
        cause: Throwable?,
        sourceLocator: SourceLocator?,
    ) : this(null, cause, sourceLocator, null)

    constructor(
        message: String?,
        sourceLocator: SourceLocator?,
        severity: Severity?,
    ) : this(message, null, sourceLocator, severity)

    val severity: Severity

    @Transient var sourceLocator: SourceLocator? = null

    val backtrace: Backtrace = Backtrace()

    init {
        this.sourceLocator = sourceLocator
        this.severity = if (severity != null) severity else Severity.ERROR
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
